### DISCUSSION TITLE
[Typescript] How to add a custom object via hook to the request variable?

### DISCUSSION THREAD

**Discussion by rubenferreira97** [fastify/fastify#3604] [Q&A]
Some of my routes need to add a custom object to the request variable before the `get/post/...` execution. All url's that have an entityId like `entities/:entityId/...` should get the entity on the req.entity. 

I couldn't make the request variable type safe with the following code:
```ts
//EntityRoutes is a plugin registered with fastify.register(Routes, { prefix: '/entities/:entityId' });
export const EntityRoutes: FastifyPluginAsync = async (fastify) => {
fastify.get<{ Params: EntityIdParamsType }>('/cars', {
    preValidation: fastify.getEntity, // fetches the entity with given Id from the DB and populates req.entity
  }, async (req, res) => {
    // Here I want to access req.entity with Entity type. 
  }
  fastify.get<{ Params: EntityIdParamsType }>('/boats', {
    preValidation: fastify.getEntity, // fetches the entity with given Id from the DB and populates req.entity
  }, async (req, res) => {
    // Here I want to access req.entity with Entity type. 
  }
}
```
I don't want to declare the entity object to all requests (or else I would declare it on fastify module).
I used preValidation hook for simplicity but if there is another approach I won't have any problem to change.

Another question, can I set a specific Params generic for all requests on this plugin? I would like to remove this code duplication `<{ Params: EntityIdParamsType }>` for every request, or even preValidation: `fastify.getEntity`.

**Comment by Eomm** [comment]
Does this question https://github.com/fastify/fastify/discussions/3597 answer this one?

**Reply by rubenferreira97** [reply]
No, I would like to specify a request type for a whole plugin (in this case a group of routes) and not only for a specific route. If possible I would like to specify the same Params for all routes as well, but this seems more unlikely and weird since it might be too "attached" to a single route.
I don't want to bind a custom object to the fastify module, because encapsulation is required (only inside this plugin).

If I read this correctly it seems this is exactly my problem:
https://github.com/fastify/fastify/discussions/2877#discussioncomment-407336

Is there any upcoming solution on fastify V4?

**Reply by Eomm** [reply]
> Is there any upcoming solution on fastify V4?

I know that the types has been rewritten: so you could try already installing the `next` branch:

```
npm i https://github.com/fastify/fastify.git#next
```

any feedback would be appreciated

**Reply by rubenferreira97** [reply]
Thanks. Are there any docs for this next branch (pre v4) version even if incomplete?

**Reply by Eomm** [reply]
No, it will be released soon
